kotlinlib/src/mill/kotlinlib/ksp/KspModule.scala

```diff
@@ -0,0 +1,156 @@
+   */
+  def kotlinSymbolProcessors: T[Agg[Dep]] = Task {
+    Agg.empty[Dep]
+  }
+
+  final def kotlinSymbolProcessorsResolved: T[Agg[PathRef]] = Task {
+    defaultResolver().resolveDeps(
+      kotlinSymbolProcessors()
+    )
+  }
+
+  override def kotlinCompilerEmbeddable: Task[Boolean] = Task { true }
+
+  /*
+   * The symbol processing plugin id
+   */
+  private val kspPluginId: String =
+    "com.google.devtools.ksp.symbol-processing"
+
+  /*
+   * The Kotlin compile task with KSP.
+   * This task should run as part of the [[generatedSources]] task to
+   * so that the generated  sources are in the [[compileClasspath]]
+   * for the main compile task.
+   */
+  private def generateSourcesWithKSP = Task {
+    val sourceFiles = sources().map(_.path)
+
+    val compileCp = compileClasspath().map(_.path).filter(os.exists)
+
+    val pluginArgs: String = kspPluginsResolved().map(_.path)
+      .mkString(",")
+
+    val xPluginArg = s"-Xplugin=$pluginArgs"
+
+    val pluginOpt = s"plugin:${kspPluginId}"
+
+    val apClasspath = kotlinSymbolProcessorsResolved().map(_.path).mkString(File.pathSeparator)
+
+    val kspProjectBasedDir = moduleDir
+    val kspOutputDir = T.dest / "generated/ksp/main"
+
+    val kspCachesDir = T.dest / "caches/main"
+
+    val pluginConfigs = Seq(
+      s"$pluginOpt:apclasspath=$apClasspath",
+      s"$pluginOpt:projectBaseDir=${kspProjectBasedDir}",
+      s"$pluginOpt:classOutputDir=${kspOutputDir / "classes"}",
+      s"$pluginOpt:javaOutputDir=${kspOutputDir / "java"}",
+      s"$pluginOpt:kotlinOutputDir=${kspOutputDir / "kotlin"}",
+      s"$pluginOpt:resourceOutputDir=${kspOutputDir / "resources"}",
+      s"$pluginOpt:kspOutputDir=${kspOutputDir}",
+      s"$pluginOpt:cachesDir=${kspCachesDir}",
+      s"$pluginOpt:incremental=true",
+      s"$pluginOpt:allWarningsAsErrors=false",
+      s"$pluginOpt:returnOkOnError=true",
+      s"$pluginOpt:mapAnnotationArgumentsInJava=false"
+    ).mkString(",")
+
+    val kspCompilerArgs = Seq(xPluginArg) ++ Seq("-P", pluginConfigs)
+
+    Task.log.info(
+      s"Running Kotlin Symbol Processing for ${sourceFiles.size} Kotlin sources to ${kspOutputDir} ..."
+    )
+
+    val compilerArgs: Seq[String] = Seq(
+      // destdir
+      Seq("-d", kspOutputDir.toString()),
+      // classpath
+      when(compileCp.iterator.nonEmpty)(
+        "-classpath",
+        compileCp.iterator.mkString(File.pathSeparator)
+      ),
+      kotlincOptions(),
+      kspCompilerArgs,
+      // parameters
+      sourceFiles.map(_.toString())
+    ).flatten
+
+    // currently if we don't delete the already generated sources
+    // several layers are problematic such as the KSP giving a FileAlreadyExists
+    // and test compilation complaining about duplicate classes
+    // TODO maybe find a better way to do this
+    os.remove.all(kspOutputDir)
+
+    kotlinWorkerTask().compile(KotlinWorkerTarget.Jvm, compilerArgs)
+
+    os.walk(kspOutputDir).filter(os.isFile).map(PathRef(_))
```

Comment:
IIUC, this collects all files under `T.dest / "generated/ksp/main"`. It's not, what we typically expect from `generatedSources`. Instead we should just return source root dirs, which is probably just `kspOutputDir`.

We already have `allSourceFiles` or `allKotlinSourceFiles` which list all individual files.

Returning files without base dirs may be even not correct in some cases. E.g. for Java files, it's required that their package declarations match their location in the directory layout.